### PR TITLE
[ROCDL] Add the global.atomic.fadd intrinsic in ROCDL

### PR DESCRIPTION
**Pull request by giuseros**
This PR adds the `global.atomic.fadd` intrinsic in ROCDL (which supports `f32` and `vector<2xf16>`)

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td [hunk 1]
```diff
@@ -516,7 +516,7 @@ def ROCDL_RawBufferAtomicCmpSwap :
 }
 
 //===---------------------------------------------------------------------===//
-// MI-100 and MI-200 buffer atomic floating point add intrinsic
+// MI-100, MI-200 and MI-300 global/buffer atomic floating point add intrinsic
```
Comment (Groverkss) [inline mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:519]: nit: at some point, we should start using architecture names (gfx9, etc.) to match llvm architecture models better

**Review by Groverkss — APPROVED** [review]

File: mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td (hunk 1 above)
Comment (krzysz00) [reply mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:519]: Yeah

File: mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td [hunk 2]
```diff
@@ -534,6 +534,19 @@ def ROCDL_RawBufferAtomicFAddOp :
   let hasCustomAssemblyFormat = 1;
 }
 
+def ROCDL_GlobalAtomicFAddOp :
+  ROCDL_Op<"global.atomic.fadd">,
+  Arguments<(ins LLVM_Type:$ptr,
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:539]: Should we go with ` LLVM_PointerInAddressSpace<1>` here to make it more obvious what's going on?

Comment (krzysz00) [reply mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:539]: and so you don't have to give the pointer type in the assembly format instead of doing a custom thing

**Review by kuhar — APPROVED** [review]

**Review by krzysz00 — APPROVED** [review]

**Comment by giuseros** [comment]
I noticed that the use case I am dealing with needs also the output from the `global.fadd` (i.e., the original value in global memory). I am converting this to draft to avoid being accidentally merged, and I will ping you guys back once I sort this out. Thanks!

**Comment by giuseros** [comment]
Done

**Review by arsenm — COMMENTED** [review]
Please just use atomicrmw fadd. I will shortly be pushing to remove the intrinsic

**Comment by giuseros** [comment]
> Please just use atomicrmw fadd. I will shortly be pushing to remove the intrinsic

Hi @arsenm , the problem is that `atomicrmw fadd` does not support vectors. So this gets translated into a cas loop which is very slow

**Comment by giuseros** [comment]
> > Please just use atomicrmw fadd. I will shortly be pushing to remove the intrinsic
> 
> Hi @arsenm , the problem is that `atomicrmw fadd` does not support vectors. So, in the case of `fp16`, this gets translated into a cas loop which is very slow

Or maybe it does?

**Comment by krzysz00** [comment]
@giuseros I wonder if it's that MLIR's wrappers around atomicrmw don't support vectors ... which seems like an extension we could do

**Comment by arsenm** [comment]
> > > Please just use atomicrmw fadd. I will shortly be pushing to remove the intrinsic
> > 
> > 
> > Hi @arsenm , the problem is that `atomicrmw fadd` does not support vectors. So, in the case of `fp16`, this gets translated into a cas loop which is very slow
> 
> Or maybe it does?

atomicrmw FP operations do since 4cb110a84f587d3c65b85d79ab6fc8aa5489fb86. I still need to implement the AMDGPU codegen changes to start using the vector instructions though (plus eventually the new metadata from #85052 will be needed

**Comment by giuseros** [comment]
Ok, given I am exactly after that vector instruction, how about we merge this PR and then we enable vector support for `atomicrmw` once it emits the vector instruction?

File: mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td [hunk 3]
```diff
@@ -165,7 +165,7 @@ def ROCDL_BallotOp :
   let summary = "Vote across thread group";
 
   let description = [{
-      Ballot provides a bit mask containing the 1-bit predicate value from each lane. 
+      Ballot provides a bit mask containing the 1-bit predicate value from each lane.
```
Comment (arsenm) [inline mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:168]: Unrelated whitespace change

Comment (giuseros) [reply mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:168]: I can add the trailing white space back, but git is warning me against this:
```
warning: 1 line adds whitespace errors.
```
My preference would be to remove the trailing whitespace, but I can remove it if you think it makes the PR cleaner.

Comment (arsenm) [reply mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:168]: Keep it out of this PR, can just precommit unrelated whitespace changes

Comment (giuseros) [reply mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td]: Done

**Comment by giuseros** [comment]
Ok, after a chat with Matthew, we agree on closing this for now and trying to emit the vectorized `atocmirmw` intrinsic. If we urgently need the feature, we will get back to this.

**Comment by arsenm** [comment]
Part 1 to start supporting the vector selection is in #94845

**Comment by arsenm** [comment]
#95393 for the LDS case, #95394 for global and flat
